// simple-react-dropdown/transpiled/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2; /* eslint-disable comma-style */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Dropdown = (_temp2 = _class = function (_React$Component) {
  _inherits(Dropdown, _React$Component);

  function Dropdown() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Dropdown);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Dropdown.__proto__ || Object.getPrototypeOf(Dropdown)).call.apply(_ref, [this].concat(args))), _this), _this.state = { isExpanded: false
    }, _this.handleDocumentClick = function (evt) {
      if (_this.mounted) {
        if (!_reactDom2.default.findDOMNode(_this).contains(evt.target)) {
          _this.setState({ isExpanded: false });
        }
      }
    }, _this.handleMouseDown = function (evt) {
      if (_this.props.disabled) return;
      if (evt.type === 'mousedown' && evt.button !== 0) return;

      _this.props.onMouseDown && _this.props.onMouseDown(evt);
      _this.setState({ isExpanded: !_this.state.isExpanded });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Dropdown, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.mounted = true;
      document.addEventListener('click', this.handleDocumentClick, false);
      document.addEventListener('touchend', this.handleDocumentClick, false);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.mounted = false;
      document.removeEventListener('click', this.handleDocumentClick, false);
      document.removeEventListener('touchend', this.handleDocumentClick, false);
    }
  }, {
    key: 'render',
    value: function render() {
      var isExpanded = this.state.isExpanded;


      return _react2.default.createElement(
        'div',
        _extends({ className: this.props.classNames.container }, this.props.containerProps),
        _react2.default.createElement(
          'div',
          _extends({
            className: this.props.classNames.trigger,
            onMouseDown: this.handleMouseDown,
            onTouchEnd: this.handleMouseDown
          }, this.props.triggerProps),
          this.props.children
        ),
        isExpanded && this.props.content
      );
    }
  }]);

  return Dropdown;
}(_react2.default.Component), _class.propTypes = { children: _react2.default.PropTypes.any.isRequired,
  classNames: _react2.default.PropTypes.object,
  containerProps: _react2.default.PropTypes.object,
  content: _react2.default.PropTypes.any.isRequired,
  disabled: _react2.default.PropTypes.bool,
  onMouseDown: _react2.default.PropTypes.func,
  triggerProps: _react2.default.PropTypes.object
}, _class.defaultProps = { classNames: {},
  containerProps: {},
  triggerProps: {}
}, _temp2);
exports.default = Dropdown;
